druid/src/box_constraints.rs

```diff
@@ -155,4 +155,218 @@ impl BoxConstraints {
 
         BoxConstraints::new(min, max)
     }
+
+    /// Test whether these constraints contain the given `Size`.
+    pub fn contains(&self, size: impl Into<Size>) -> bool {
+        let size = size.into();
+        (self.min.width <= size.width && size.width <= self.max.width)
+            && (self.min.height <= size.height && size.height <= self.max.height)
+    }
+
+    /// Find the `Size` within these `BoxConstraint`s that minimises the difference between the
+    /// returned `Size`'s aspect ratio and `aspect_ratio` An aspect ratio is defined as
+    /// `height / width`.
+    ///
+    /// If multiple `Size`s give the optimal `aspect_ratio`, then the one with the `width` nearest
+    /// the supplied width will be used. Specifically, if `width == 0.0` then the smallest possible
+    /// `Size` will be chosen, and likewise if `width == f64::INFINITY`, then the largest `Size`
+    /// will be chosen.
+    ///
+    /// Use this function when maintaining an aspect ratio is more important than minimizing the
+    /// distance between input and output size width and height.
+    ///
+    /// TODO should aspect_ratio be restricted to 0 < ar < infinity and panic otherwise? Same for
```

Comment:
I think it's reasonable to at least `abs()` aspect ratio / width. I think it should also be an invariant that constraints are never negative, and we should perhaps be enforcing this on creation, but that would be a separate PR.